### ISSUE TITLE
Overloading milestone hooks

### ISSUE THREAD

**Issue by carera** [dchester/epilogue#155]
Hi,
I am trying to overload default behavior of - for example - POST method on an endpoint:
`resource.create.write((req, res, ctx) => {
  // my custom DB insert here
});`
But epilogue only registers this hook after the default one, so that both the default insert and my custom one are executed during request, resulting in an unwanted behavior.

Is there a way of overriding/overloading/rewriting/cancelling the default hook, instead of just adding custom ones?

**Comment by mbroadst** [comment]
you should be able to do something like this:
```
resource.create.write.before((req, res, ctx) => {
  // do something
  return ctx.skip(); // skip the rest of this milestone
});
```

**Comment by carera** [comment]
Oh, thanks, but doesn't this fundamentally undermine the epilogue philosophy of having granular control over request lifecycle? All customised behaviour will then be defined in `.before` hooks which does not make much sense from the logical point of view.

**Comment by mbroadst** [comment]
@carera the milestones are implemented as a stack, are you sure that your code is being run after the default implementation when doing this:

```
resource.create.write((req, res, ctx) => {
  // my custom DB insert here
  ctx.skip();   // skips the default impl
});
```

If thats not the case, a test case would be appreciated and we can get on working on this as if it were a bug.

I think the epilogue philosophy is to be able to do whatever you need to do, and hopefully the milestone design is flexible enough to accommodate that. If you really look at epilogue its a "milestone system" with a bunch of sane defaults baked in for restful operations.

**Comment by carera** [comment]
@mbroadst Thank you for reaching back, I have created a PR with the testcase: #https://github.com/dchester/epilogue/pull/156

the thing is fairly simple, the following fails on duplicate mail attribute entry 
`Unhandled rejection SequelizeUniqueConstraintError: must be unique`
`
    it('should overload create.write milestone with custom one', function(done) {
      test.userResource.create.write(function (req, res, context) {
        test.models.User.create({
          username: req.body.username,
          email: req.body.email
        }).then(function(result) {
          context.instance = result;
          return context.skip();
        });
      });
      request.post({
        url: test.baseUrl + '/users',
        json: { username: 'arthur', email: 'arthur@gmail.com' }
      }, function(error, response, body) {
        expect(response.statusCode).to.equal(201);
        expect(response.headers.location).to.match(/\/users\/\d+/);
        done();
      });
    });
`

**Comment by mbroadst** [comment]
@carera oops sorry I've been reading this incorrectly from the get-go (early morning here). This behavior is expected, and the best practice for overriding the default `write` milestone in this case is to implement the `.write.before` and use `ctx.skip()` to skip that milestone.

**Comment by carera** [comment]
@mbroadst thanks, this still however logically does not make sense. Shouldn't the actual data insertion happen in the `action` of that milestone? I would still like to use `before` and `after` hooks for pre- a post- processing. Now, it will just be either a single spaghetti code within a `before` hook, or a bloat of `.before` hooks, run one after another, if you know what I mean. As I was saying in https://github.com/dchester/epilogue/issues/155#issuecomment-209887778, a project with a slightly more customized behaviour will end up bloated with `.before` 'hacks'

**Comment by mbroadst** [comment]
@carera yeah I understand, but the only way to change that at this point is to reverse the order in which milestones are evaluated which could be potentially disastrous for existing users (unless you have an alternative proposal).  I do agree it makes more sense, but this is legacy behavior from the initial release of the module afaik, but I guess we could flip it around and push a 1.0.0 release?

**Comment by carera** [comment]
@mbroadst see https://github.com/dchester/epilogue/issues/155#issuecomment-209887778
A possible way would be to give the developer an option to override, or to skip the default hook:
`resource.action.milestone.overrideAction(…)`
or
`resource.action.milestone(fn, override)`
or maybe directly in the resource definition

**Comment by mbroadst** [comment]
@carera basically the change boils down to changing this line: https://github.com/dchester/epilogue/blob/master/lib/Controllers/base.js#L207

from `this[name].push(callback);` to `this[name].unshift(callback);`

**Comment by carera** [comment]
@mbroadst I kind of like your proposal, a smart and efficient way of working this out, but as you are saying, the potential effects on existing solutions using epilogue could be disastrous. Also, a developer would have to keep in mind that multiple declarations of the same hooks would be then run bottom-up

**Comment by carera** [comment]
But even that way, if you call `context.skip()` in the `resource.create.write`, the `.after` hook is still skipped afaik

**Comment by mbroadst** [comment]
@carera hence the major version bump. Users would have to opt-in to this new functionality, and would therefore know what was happening. I think the other potential benefit is that milestones would be operate in the same fashion as express (connect) middleware, or restify, so as to keep the convention going.

Regarding your second comment: the way `context.skip()` works is to skip the rest of the milestone including any before or after hooks. Again this is just a design decision that was made early on.

**Comment by mbroadst** [comment]
oops looks like I was wrong, they do in fact process the routes on a stack like we do.. back to the drawing board

**Comment by carera** [comment]
@mbroadst In express you build the stack yourself. Here in Epilogue you have the default stack pre-built. My idea is to intercept any of the hooks and overload them, not really stack more of them or skip them.
For example, in `read` action, you have the following life-cycle:
```
  .start.before
  .start.action
  .start.after
  
  .auth.before
  .auth.action
  .auth.after
  
  .fetch.before
  .fetch.action
  .fetch.after
  
  .send.before
  .send.action
  .send.after
  
  .complete.before
  .complete.action
  .complete.after
```
And i just want to change the way the actual "find one" query is executed, so i intercept the .fetch.action with my custom function, and the life-cycle continues on

**Comment by mbroadst** [comment]
@carera Okay yeah I see what you're saying. I dig your idea from above (with potentially slight modification) for more "low level" overriding of milestones:

we would modify the existing `milestone` controller method (https://github.com/dchester/epilogue/blob/master/lib/Controllers/base.js#L197) to be:
```
Controller.prototype.milestone = function(name, options, callback) { ...
```

allowing for an `options.override` which will optionally completely replace the milestone instead of appending to it.

So you would have something like:
```
userResource.create.milestone('write', { override: true }, (req, res, ctx) => {
  // your stuff
  return ctx.[whateverYouWantToDo]();
});
```

Thoughts?

**Comment by carera** [comment]
I like it, I however in no way want to force you to change the fundamentals of Epilogue, "just because some script kiddie has a different use case". Maybe i misunderstood the philosophy completely. But then again even the documentation says
`Controller actions in turn have hooks for **setting and overriding behavior** at each step of the request. `
or
```
action: function(req, res, context) {
        // **change behavior** of actually writing the data
        return context.continue;
      },
```
When I went for help with this to a development team next-door, who has been using epilogue for a while now, it was only after I pointed this issue out that they figured that all their as-they-thought-overriden `resource.list.fetch` was executing both default and their custom queries on each request.

Before you jump into changing the guts of the framework, I would like to know what was the reason behind designing the behavior this way, or what was the use case of it?

**Comment by mbroadst** [comment]
@carera to be quite honest with you I didn't design it, I just took over maintenance of this module a few months after it was put out while I was using it for a project.. and here I am a few years later :smile: (caveat: I haven't actually used epilogue in a number of years)

That documentation I believe was written by contributors along the way as they navigated troubles with the implementation, and here we find ourselves today with a mostly working solution.  The advice generally given for your case has been to use the `.before` hooks to modify the behavior, but I appreciate the need for more fine-grained behavior (in e.g. your use case). 

To that end, I don't intend to change the internals of epilogue but rather continue my "stewardship" of the module. I'm here to do maintenance on the module as-needed, field questions, process pull requests and occasionally release.  If you want to submit a patch to enable this behavior it's more than welcome (as long as tests are included), I think the solution is a good one, falls in line with the tenants of the project up to this point, and to boot it isn't even a breaking change.

**Comment by carera** [comment]
I completely understand, and you are doing good job! :)
Anyway I did toy with the modification and it's fairly easy (see work-in-progress commit: https://github.com/dchester/epilogue/compare/master...carera:milestone_overloading)
The problem with this I see is that this gives sort of an "inconsistency freedom" to the user. Imagine defining multiple hooks:
```
resource.create.write(function (req, res, context) {...}
resource.create.write({override: true}, function (req, res, context) {...}
resource.create.write({override: true}, function (req, res, context) {...}
resource.create.write(function (req, res, context) {...}
```
Then the final stack will only have certain hooks (last two in this case). What if instead the user had to somehow explicitly state, that he wants to extend the default functionality, instead of overriding it (same way as with `this.super()` in OO inheritance)?

**Comment by mbroadst** [comment]
Well of course it does, you're allowing them to override the target completely.  At a certain point you need to trust the user, or rather that you've given them the tools to get the job done. 

To what degree do you mean "extend the default functionality"? How does simply defining a new milestone for an action with today's code not extend the default functionality?

**Comment by chinayuans** [comment]
hi, 
   I think carera's advice is very useful. I eagerly need it.
